cheroot/test/conftest.py

```diff
@@ -0,0 +1,54 @@
+import threading
+import time
+
+import pytest
+
+import cheroot.server
+import cheroot.wsgi
+
+EPHEMERAL_PORT = 0
+NO_INTERFACE = None  # Using this or '' will cause an exception
+ANY_INTERFACE_IPV4 = '0.0.0.0'
+ANY_INTERFACE_IPV6 = '::'
+
+config = {
+    'bind_addr': (NO_INTERFACE, EPHEMERAL_PORT),
+    'wsgi_app': None,
+}
+
+
+def cheroot_server(server_factory):
+    conf = config.copy()
+    bind_port = conf.pop('bind_addr')[-1]
+
+    for interface in ANY_INTERFACE_IPV6, ANY_INTERFACE_IPV4:
```

Comment:
@jaraco turns out, if Linux box supports dual stack it will bind to `::`, but also be available as `0.0.0.0:sameport`. And this [raises `OSError: [Errno 99] Cannot assign requested address` on the client side](https://travis-ci.org/cherrypy/cheroot/jobs/320644912#L593).

One can emulate this on linux using `sudo sysctl net.ipv6.conf.all.disable_ipv6=1`, which will disable IPv6 addresses on interfaces, but `socket.has_ipv6` **will still remain `True`**, because *kernel is built with IPv6 support*.

@jaraco do you have any ideas on how can we treat this?

Reply:
I've [previously posted that Python should provide a routine for binding in dual-stack environments](https://bugs.python.org/issue25667), but in that ticket, I reference back to the CherryPy implementation, which is now part of this project.

You say that one can emulate this on Linux by disabling IPv6 - is that what's happening in Travis?

Reply:
Looks like there are recent, emergent [issues with IPv6 on Travis](https://github.com/travis-ci/travis-ci/issues/8891), and that's probably what's happening - IPv6 is only partially available. Perhaps this routine for creating a server should check that a connection can be established on IPv6 (not just that a connection can be bound on IPv6) when setting up the server... or maybe in the server_client, fallback to IPv4 as a workaround for IPv6-broken environments like Travis.

Reply:
I've seen this kind of issue lots of times in Travis and don't want to rely on its state.

Reply:
I was hoping that there's some clever syscall capable of telling us whether IPv6 is disabled in sysctl

Reply:
> Perhaps this routine for creating a server should check that a connection can be established on IPv6 (not just that a connection can be bound on IPv6) when setting up the server... or maybe in the server_client, fallback to IPv4 as a workaround for IPv6-broken environments like Travis.

Yes, I was thinking about such options, but I still don't have a clear vision w/o hacks on how to do this check sanely. Technically, I can connect to the server if I replace `::1` with `127.0.0.1`, but I don't like such hardcoding. On one hand, I'd leave `bind_addr` representation as it comes from the OS in the server object, because it's how things work in such envs, on the other hand, I'd like to see some limitations/flags that it also listens to IPv4 on the same port to allow client decide on the best protocol available from its perspective.

Reply:
New info, to be clear:

According to https://wiki.archlinux.org/index.php/IPv6#Disable_functionality, there's `ipv6.disable=1` and `ipv6.disable_ipv6=1`. The first one is destined to disable the whole IPv6 network stack in kernel (also achieved by `rmmod` if it's been built as a separate module, rather than built-in to kernel itself). **The second one will keep the IPv6 stack functional but will not assign IPv6 addresses to network devices.** And this is exactly what's happening.

So I think it should be handled on the client side.